### PR TITLE
Dev Console Improved Parsing

### PR DESCRIPTION
**Pull request by BraedonWooding**
This is a small set of changes, it just is focused at removing class files and making dev console run smoother.  It removes the help function and adds another string called 'DetailedDescription' (@koosemose suggested this), this is printed whenever `Command ?` is called (note if it doesn't exist it'll just print the normal description).  

The system now has a lot simpler definition for internal commands (non-mod based ones), you can just add an attribute in format `[Command(params string[] tags)]` there are a few properties that you can access (description, detailedDescription, and title in the format of `title = "MyTitle"`...).  Again you can't just do this in the modded C#!!!!  You have to use xml/json for that (to keep it inline with the LUA) in the future I will add support to just do an attribute in C# but not currently.

The system has removed the old method of parsing arguments, it used to get the type then access all its constructors and call the one that matches the types of parameters required and so on... for structs/classes and for primitives it had a case switch situation.  Now you can just add a simple attribute to any function that takes in a string and returns an object, it is in format `[Parser(typeof(MyType))]`.  Again as above only works in our main assembly, while this does 'limit functionality' (you need a parser for every type wanted and that includes custom classes), you still had to have a case for every primitive so it really only adds a small amount of 'bloat' for vectors currently (the only non-primitive) and eventually a few others.  Most parse functions are only two lines (with a few three liners) and if unity ever brings in the new C# version with the inline out variables they can be made into one liners (with a few two liners).  So if you have a better idea for them please put them below!

I have no intentions of allowing LUA to either parse the string representations into actual object types or have an attribute form of commands rather than the JSON/XML file, the reason being that LUA has different data types and meanings that can't be carried over (i.e. ushort/uint16 doesn't exist in LUA), also LUA doesn't support attributes (though maybe we have an init function to allow a function call though that is basically what the JSON/XML file is...?).  

## Removes
- All 3 interfaces
- The error class
- A lot of the old parsing code
- A lot of the speciality code

### CODE REVIEW

**Comment by koosemose** [comment]
In theory, Lua shouldn't need to do too much in the way of parsing into types, for any core type (int, bool, string, etc...) it typically just does what you want (treats it as whatever it is you're trying to use it as. The only thing it'd have issues with is objects, but most (all?) of the time, those are going to be standard PP objects (furniture and co.) which could be (if not already) handled in core. And then of course there's arrays and other collections, and for those, as long as they can be done in a way that passes the collection along, most (all?) collections should be translated to tables (which lua uses to cover all of those). 

In short, lua *shouldn't* need any ability to parse, because most anything it can be passed that would make sense would likely be either already handled by the system or not available to anything. Anything that would need special parsing (like a new object perhaps?) would need C# code anyways.

**Comment by BraedonWooding** [comment]
The parsing is to get a string representation into the proper type ie convert "1" to 1, "(0, 2, 3)" to a vector of 0, 2, 3.  Then they are passed in during an invoke call (which takes in an object array).  Explained it poorly oops.

Basically it comes down to LUA having different datatypes so it not being good at converting string representations to the C# variant.  And things like furniture wouldn't really be passed in as a parameter to a dev console command, you would pass in a furniture name though I guess you could build a parser that could take a representation and turn it into an object for dev console functions.

**Comment by koosemose** [comment]
Yeah... basically what I was saying is that Lua shouldn't need any complex parsing, if it just gets a string, it'll do what's most often wanted... it shouldn't have to worry about converting to C# types, since if you (as a modder) need a C# type, you're probably already doing C# modding... except for standard things (such as your example of a vector, I'm assuming from a modder perspective they won't actually have to do any special handling of such a common type as a vector3

**Comment by BraedonWooding** [comment]
Yep that's what the parsers are for, they make it so modders don't ever have to do custom parsing, they just step in right before and convert them for you basically.  An example flow chart would be this: User Enters Command -> Process Name and find target command -> Find parameters and types -> Convert the string to the type that function wants -> Call the function with the requested parameters given by the user.

**Comment by BraedonWooding** [comment]
This is all done (with obviously regard to it passing travis and your checks/lookover)

**Comment by koosemose** [comment]
So am I understanding correctly that creating a core command now is simply a matter of putting an attribute on a function no matter where it is? Can you point out a few locations where core commands are created so I can look over ones in actual use (I find it much easier to process how things work when looking at things in actual use.)

**Comment by BraedonWooding** [comment]
In DevConsole.cs (the main file) search for Help, Clear, and AllTags (around line 600).  They are currently the only commands that use the attribute.  And yes it is just a single attribute.  Same with the parsers which are right below those commands which are also just a single attribute, again the parsers are used to parse the string representation into the correct variant.  All modded variations still have to use the old system of the xml/json file (which all the other commands use) however, when I get some time I'll add it so you don't have to and can use the attribute for commands and add custom parsers (only for C# commands since they are the only ones that can have attributes and this means LUA will still require the xml/json file).

**Comment by koosemose** [comment]
It might be wise of us to require (either through technical means or just policing any PRs) core commands to all be in the same place (even if that leads to them just calling a function elsewhere) just for the sake of ease of maintenance (just imagine some point in the future you change some element of how commands are specified, and now, rather than just scanning through a list of commands you have to search through all files... not impossible, but at least an annoyance, and possibly more extra added layers of complexity if it is someone other than yourself doing it.)

**Comment by BraedonWooding** [comment]
Yeh I agree.  Should I create a new file just to house the 3 system commands (since the other ones are not really 'system'/core they are stuff like increase temperature, kill this guy, status of the game and so on...) and all of the parsers (there are quite a few)?

**Comment by BraedonWooding** [comment]
Yes I completely agree, should I create a new file to house all the system commands and parsers?  The modded ones are all in that mod file which I think is good though some of them one could argue should be moved out?

**Comment by koosemose** [comment]
A new file would probably be nice and neat, if it wouldn't add any excess complexity. And in most cases commands aren't performance critical, so it doesn't matter too much if they're core or not (though if there are any exceptions they should of course be moved), and we should have a selection of "modded" commands left to serve as examples.

**Comment by BraedonWooding** [comment]
Nah it wouldn't, would just be a static sealed class (purely just to indicate that it is just for holding those commands the commands don't require a static nor sealed class just themselves to be static).  And actually modded commands are faster, with modded commands they are compiled and then run whereas with internals I have to perform reflection (pretty much the same speed but compiled is somewhat 20% faster in some cases).  And yeh as you said they aren't performance critical which is why I didn't do delegate tricks or stuff since that would just add a lot of complexity for not a great performance gain especially in an area that isn't going to be run much or needs to be fast.

Personally I like a large selection of modded stuff as examples which is why all the 'generic options' for the settings menu are implemented as mods (and when we have multi-mods introduced they will be put in a utilities mod since at one point I'm sure someone will want to disable core, and to remove those options is odd) regardless, I'll go through and check that well one they are all working as intended and two are in the right place.

**Comment by BraedonWooding** [comment]
Due to PP not being a fork and that stuffing up one or two of my PRs going to close and reopen the 'new PR'
